### ISSUE TITLE
Java: pairing does not work multithreaded

### ISSUE THREAD

**Issue by manudrijvers** [milagro-crypto/amcl#43]
Similar to what @samueleAndreoli fixed in f88274f17bcfbce626da94962fefa9c9507a7241, there is a multithreading problem in computing a pairing. We make a = g2^5, b = g1, c = g2, d = g1^5, and check that e(a, b) = e(c, d). This holds if we run one thread, but with two threads, it starts failing. The following test demonstrates the problem. 

```
public class MultithreadAmcl {
    @Test
    public void testAmclThreads() throws InterruptedException {
        BIG gx = new BIG(ROM.CURVE_Gx);
        BIG gy = new BIG(ROM.CURVE_Gy);
        ECP genG1 = new ECP(gx, gy);

        ECP2 a = new ECP2();
        ECP b = new ECP();

        a.copy(genG2);
        a = a.mul(new BIG(5));

        b.copy(genG1);

        ECP2 c = new ECP2();
        ECP d = new ECP();
        c.copy(genG2);

        d.copy(genG1);
        d = d.mul(new BIG(5));


        VerifyAmcl t1 = new VerifyAmcl(a, b, c, d, "T1");
        VerifyAmcl t2 = new VerifyAmcl(a, b, c, d, "T2");
        List<VerifyAmcl> runners = new ArrayList<>();
        runners.add(t1);
        runners.add(t2);
        for (VerifyAmcl runner: runners) {
            runner.start();
        }

        for (VerifyAmcl runner: runners) {
            runner.thr.join();
        }

    }

    private class VerifyAmcl implements Runnable {
        private ECP b, d;
        private ECP2 a, c;
        private Thread thr;
        private String name;

        public VerifyAmcl(ECP2 a, ECP b, ECP2 c, ECP d, String name) {
            this.a = a;
            this.b = b;
            this.c = c;
            this.d = d;
            this.name = name;
        }

        public void invoke() {
            assertTrue(PAIR.fexp(PAIR.ate(a, b)).equals(PAIR.fexp(PAIR.ate(c, d))));
        }

        public void start () {
            System.out.println("Starting " +  name );
            if (thr == null) {
                thr = new Thread (this, name);
                thr.start ();
            }
        }
        @Override
        public void run() {
            for(int index = 0; index < 100; index++) {
                invoke();
            }
            System.out.println("Thread " +  name + " exiting.");
        }
    }
}
```

**Comment by mcarrickscott** [comment]
Hello Manu,

I just love debugging multi-threaded code :)

But I think I nailed it.

Mike

**Comment by manudrijvers** [comment]
Don't we all :). Thanks for the quick update Mike, we're gonna check this out and let you know how it goes!

**Comment by dubovitskaya** [comment]
Hello Mike, 

this is Rafa and Maria - we work with Manu on Idemix for Hyperledger Fabric that uses the amcl library. 

Many thanks for such a quick fix for the pairing! 
Unfortunately, after we tried our high-level multithread test again, we found other issues with multithread safety in amcl. 
To make sure this is not our fault (not from a high-level code that uses amcl) we created a fork of amcl for one curve that we use and patched the amcl library for this one curve (BIG, ECP, ECP2, FP classes) + added multithread tests for other operations (G1Mul, etc.). 

This is the link to the fork with tests and patches (you can see also the commit history with the changes we made):
https://github.com/rafaeltm/amcl-java-fp256bn

This version passes our high-level multithread test.

Could you please take a look at the proposed patches and let us know if they make sense to you and  could be propagated to the entire library (so that they work for other curves) and the version in the central maven repo can be updated? 

Many thanks and best regards,
Rafa and Maria

**Comment by mcarrickscott** [comment]
Hello Maria and Rafa,

Thanks for the feedback. I had a look. It seems the changes you want are
just to the files BIG/ECP/ECP2 and FP.

The changes fall into 4 categories.

1. In ECP/ECP2 you would like a new constructor. Good idea

2. You would like Arrays.CopyOf used (more efficient?).

3. You would like StringBuilder used (not sure why exactly)??

Can I assume that these changes, while all good ideas, are not relevant to
the multi-threading problems you have been having?

Which brings me to the main problem as I see it. AMCL sometimes internally
tidies up its internal representation of big numbers. If two or more
threads have read-only access to the same value, then this behavior will
cause a problem. I am not sure if you have caught all the instances where
this might happen. It would seem prudent to me for each thread to create
its own copy of such read-only data. Or maybe I am missing something (I am
far from being a Java expert, and certainly not in multi-threaded Java!).

Anyway I am happy to make all of the changes indicated as patches in those
files, when you confirm that I have understood the full nature of the
problem.

Mike

**Comment by dubovitskaya** [comment]
Hi Mike, 

Thanks a lot for your fast answer!

> Can I assume that these changes, while all good ideas, are not relevant to the multi-threading problems you have been having?

Indeed, we applied two types of changes – the ones that we think are fixing problems with the multithread issues and also some style changes (like StringBuilder for concatenation) – sorry if that was confusing, we should have removed the second ones to avoid the confusion.

We think that the changes that fall into the first category should address (some of) the multithread issues. For example, adding the constructors are necessary to provide object copies, using Arrays.CopyOf() ensures that we are working with the object copies when copying.

Basically, addressing the ‘copies’ issues (avoiding the use of the original objects) in the following classes & methods helped us fix the memory sharing problems - at least the ones that showed up in our multithread test: 

1. BIG:   pmul, pxmul, smul, mul, sqr, monty, minus, tobytearray, modmul, modsqr, modneg, powmod, plus, add, and mod.
2. ECP: mul2
3. ECP2: copy constructor
4. FP: equals

We do understand that the multithread debugging is not the easiest thing and that the proposed fixes might not solve all potential multithread issues. 

Could you please let us know what you think? We would happy to help or further investigate the problem if needed. 

Many thanks and best regards,
Rafa and Maria

**Comment by mcarrickscott** [comment]
One thing help would be explicit instructions as to how to run your tests.

I am not very familiar with Maven..

Mike

**Comment by dubovitskaya** [comment]
sure, just sent you the instructions by email.

Many thanks and best regards,
Rafa and Maria

**Comment by mcarrickscott** [comment]
OK, I see the problem. Give me a few days.

Mike

**Comment by mcarrickscott** [comment]
OK, try that.

I only made those changes that might impact the multi-threading issue.

Let me know exactly what tests (if any) that it now fails, so that I can
reproduce the errors here. I figured out how to run your tests - of the
original group I was only able to detect errors with Mult2Test


Mike

**Comment by dubovitskaya** [comment]
Hi Mike,

Many thanks! 

Did you try to run the high-level test that I sent you by email? IdemixTest?
With the update it started failing even with one thread for us.

Best regards,
Maria

**Comment by mcarrickscott** [comment]
Unfortunately it bombs out with this error

FATAL ERROR in native method: processing of -javaagent failed
Aborted (core dumped)

Results :

Tests run: 0, Failures: 0, Errors: 0, Skipped: 0

[INFO]
------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO]
------------------------------------------------------------------------
[INFO] Total time: 24.388 s
[INFO] Finished at: 2018-08-28T16:26:18Z
[INFO]
------------------------------------------------------------------------
[ERROR] Failed to execute goal
org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test)
on project fabric-sdk-java: Execution default-test of goal
org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test failed: The
forked VM terminated without properly saying goodbye. VM crash or
System.exit called?
[ERROR] Command was /bin/sh -c cd /home/shamus/fabric-sdk-java &&
/usr/lib/jvm/java-11-openjdk-amd64/bin/java
'-javaagent:/home/shamus/.m2/repository/org/jacoco/org.jacoco.agent/0.7.9/org.jacoco.agent-0.7.9-runtime.jar=destfile=/home/shamus/fabric-sdk-java/target/coverage-reports/jacoco-ut.exec,excludes=**/org/hyperledger/fabric/protos/**'
-jar
/home/shamus/fabric-sdk-java/target/surefire/surefirebooter8710195638089277920.jar
/home/shamus/fabric-sdk-java/target/surefire/surefire17978393060866462509tmp
/home/shamus/fabric-sdk-java/target/surefire/surefire_015796437197325954298tmp
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e
switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions,
please read the following articles:
[ERROR] [Help 1]
http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException


However I did look at some of the code you are using. It may be a different
problem now. I will look at it some more...

Mike

**Comment by mcarrickscott** [comment]
There was a  bug in my last commit.

Just updated it again..

Mike

**Comment by mcarrickscott** [comment]
One problem I am having is understanding how some of the changes you have
suggested could possibly have any affect.

For example how could this:-

/* return this+x */
    public BIG plus(BIG x1) {
        BIG x = new BIG(x1);
        BIG s = new BIG(0);
        for (int i = 0; i < NLEN; i++)
            s.w[i] = w[i] + x.w[i];
        return s;
    }

be any different from this:-


/* return this+x */
public BIG plus(BIG x) {
BIG s=new BIG(0);
for (int i=0;i<NLEN;i++)
s.w[i]=w[i]+x.w[i];
return s;
}

How can making an internal copy of x have any impact? The parameter x/x1 is
read-only, and never written to by the function, so multiple threads
calling the function cannot have any unfortunate interaction.

Of maybe I am missing something...

I am willing to make these changes, but would love some kind of explanation!


Mike

**Comment by dubovitskaya** [comment]
Hi Mike,

Many-many thanks for the latest fix! We tested it extensively with our multithread test and it works. 

Regarding your question about our changes – we were just trying to make things work so we might have been extra “paranoid” about some of the changes. Please discard them if you do not find them necessary. 
Our code was not a PR request – more a working example and we hope that it helped in some way.
Many thanks again for providing such a quick update!

Could you please update the central maven repo with your fix as well so that we can revendor it properly in Fabric?

We have another question regarding the Go version – I will send you an email about it. 
Could you please have a look? 

Many thanks again!

Best regards,
Rafa and Maria

**Comment by kealan** [comment]
Hi Rafa and Maria

We have updated the milagro-crypto-java code with the latest changes and pushed to Maven Central repo - version 0.4.0

**Comment by rafaeltm** [comment]
Hi Kealan and Mike,

Thank you so much for your work!

We're glad to help you improving this lib!

Best regards,
Maria and Rafa

**Comment by manudrijvers** [comment]
Thanks for your help Mike and Kealan!
